lib/warpath/expression.ex

```diff
@@ -27,6 +27,7 @@ defmodule Warpath.Expression do
   @type filter :: {:filter, has_property | {operator | fun, term}}
   @type scan :: {:scan, property | wildcard | filter | array_indexes}
   @type token :: root | dot_access | wildcard | array_indexes | filter | scan
+  @type t :: list(token)
```

Comment:
There is a opportunity to improve this api, instead of return `list(tokens)` `Expression.compile` could produce a `%Expression{tokens: tokens}` struct, this will keep the extensibility of Expression module flexible to the future.
Of course this will generate more work to be done, because the change needed in the `ExpressionTest` and `Warpath` modules.
What to you think?

Reply:
Yes, absolutely. I think having an explicit struct/type to pattern-match on is always a good idea 👍

Reply:
@jonasschmidt The parser refactory have been merged, we can advance with this feature, could your implement the suggestion above?

Reply:
Sorry, I was a bit busy with other work. I will have another look today.

Reply:
:+1:  Great, if you have any question, ask here that I will try to help you.